### PR TITLE
Refactoring Span/ReadOnlySpan Extensions into partial classes

### PR DESCRIPTION
**Pull request by ahsonkhan**
- Separating shared and platform specific implementations

The CI should pass after the following CoreCLR change is merged:
https://github.com/dotnet/coreclr/pull/9182

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by ahsonkhan** [comment]
cc @KrzysztofCwalina, @kouvel

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

File: src/System.Memory/src/System/SpanExtensions.cs [hunk 1]
```diff
@@ -13,105 +13,102 @@ namespace System
...
-        /// <exception cref="System.OverflowException">
-        /// Thrown if the Length property of the new Span would exceed Int32.MaxValue.
-        /// </exception>
+        /// <param name="span">The span to search.</param>
+        /// <param name="value">The value to search for.</param>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static Span<byte> AsBytes<T>(this Span<T> source)
-            where T : struct
+        public static int IndexOf<T>(this Span<T> span, T value)
+            where T:struct, IEquatable<T>
```
Comment (KrzysztofCwalina) [inline src/System.Memory/src/System/SpanExtensions.cs]: Don't we already have this overload in Span? i.e. should it simply be removed from extensions? Same for other indexof overloads.

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

File: src/System.Memory/src/System/SpanExtensions.cs (hunk 1 above)
Comment (ahsonkhan) [inline src/System.Memory/src/System/SpanExtensions.cs]: Span does not contain these IndexOf overloads. They only exists in the SpanExtensions class. Removing them here would break the contract defined in the System.Memory ref. Where do we have duplicates of these overloads?

File: src/System.Memory/src/System.Memory.csproj [hunk 2]
```diff
@@ -27,12 +24,22 @@
     <Compile Include="System\SpanHelpers.Clear.cs" />
     <Compile Include="System\ThrowHelper.cs" />
   </ItemGroup>
-  <ItemGroup>
+  <ItemGroup Condition="'$(IsPartialFacadeAssembly)' != 'true'">
```
Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: Could we just remove the netstandard1.0 build in Configuration.props, and remove any files that are specific to that build? @joperezr, is there any reason to build for older frameworks anymore?

If we should continue to build for netstandard1.0, I suggest using a condition based on the TargetGroup so that it's clear which runtime which items are for.

Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: Since it looks like you're also enabling CoreCLR's span with this change, could you please also remove the restriction in Configuration.props that builds for Windows only? This would work fine I think:
```xml
    <BuildConfigurations>
      netcoreapp;
    </BuildConfigurations>
```

And can remove any windows-specific stuff in the csproj.

File: src/System.Memory/src/System.Memory.csproj [hunk 3]
```diff
@@ -27,12 +24,22 @@
...
     <ProjectReference Include="..\..\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj" />
     <Reference Include="System.Diagnostics.Debug" />
     <Reference Include="System.Reflection" />
     <Reference Include="System.Resources.ResourceManager" />
     <Reference Include="System.Runtime" />
   </ItemGroup>
+  <ItemGroup Condition="'$(IsPartialFacadeAssembly)' == 'true'">
+    <ProjectReference Include="..\..\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj" />
+    <ReferenceFromRuntime Include="System.Private.CoreLib" />
+    <ReferenceFromRuntime Include="System.Runtime" />
```
Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: If this is targeting System.Runtime, it should not be necessary to also target System.Private.CoreLib. Are there any types necessary from CoreLib that are not in CoreFX's contracts?

Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: My understanding is that it's not good to target both because there would be name conflicts. Most assemblies target either only CoreLib, or only other CoreFX contracts.

File: src/System.Memory/ref/System.Memory.cs [hunk 4]
```diff
@@ -91,13 +72,20 @@ public partial struct Span<T>
         public bool TryCopyTo(System.Span<T> destination) { throw null; }
     }
 
-    public static class SpanExtensions
+    public static class Span
```
Comment (kouvel) [inline src/System.Memory/ref/System.Memory.cs]: As this is effectively a breaking change (though only for the location of extension methods, which may not be a big deal), we should probably bump the assembly version to 4.1.0.0 in the ref and src projects.

Comment (kouvel) [inline src/System.Memory/ref/System.Memory.cs]: Actually I believe the assembly version comes from System.###\dir.props, not from the csproj files, as done in other places. Probably should add that file and remove the AssemblyVersion property from the two csproj files.

File: src/System.Memory/src/System.Memory.csproj (hunk 3 above)
Comment (mellinoe) [inline src/System.Memory/src/System.Memory.csproj]: Partial facades can and often need to reference both System.Private.CoreLib and System.Runtime. Keep in mind that this is the implementation assembly for System.Runtime, so there are no type conflicts (System.Runtime is itself built against System.Private.CoreLib in this case).

File: src/System.Memory/ref/System.Memory.cs (hunk 4 above)
Comment (mellinoe) [inline src/System.Memory/ref/System.Memory.cs]: This assembly has never been shipped stable, as far as I know, so breaking changes are not a concern.

File: src/System.Memory/src/System.Memory.csproj (hunk 2 above)
Comment (mellinoe) [inline src/System.Memory/src/System.Memory.csproj]: This is an "artificial" restriction that needs to be kept for now. This library relies on System.Runtime.CompilerServices.Unsafe, which cannot be built outside of Windows right now.

File: src/System.Memory/src/System.Memory.csproj (hunk 3 above)
Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: I see, makes sense, thanks

File: src/System.Memory/src/System.Memory.csproj (hunk 2 above)
Comment (mellinoe) [inline src/System.Memory/src/System.Memory.csproj]: The netstandard configuration is needed because there are two distinct assemblies being generated: the "fast span assembly" with references to System.Private.CoreLib's private implementation, and the "slow span assembly" with no runtime dependencies.

Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: What would the slow span assembly be used for?

Comment (mellinoe) [inline src/System.Memory/src/System.Memory.csproj]: The fast span assembly can only be used in .NET Core 2.0 (`netcoreapp2.0`), so the slow span assembly would be used everywhere else.

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

File: src/System.Memory/src/System.Memory.csproj [hunk 5]
```diff
@@ -16,9 +16,6 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netcoreapp-Windows_NT-Release|AnyCPU'" />
   <ItemGroup>
     <Compile Include="System\Pinnable.cs" />
```
Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: Shouldn't be necessary to include this for the netcoreapp build

File: src/System.Memory/src/System.Memory.csproj (hunk 2 above)
Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: Are we planning on shipping the slow span assembly in the future for some runtime other than for .NET Core 2.0? What I'm not following is that we're building it currently for netstandard1.0 - are we planning on shipping the assembly again for that? If not, why build it? If we are planning to ship it for other runtimes like UAP or CoreRT then I would imagine we would build for those configurations, not for netstandard1.0.

Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: Is it possible to use CoreLib's Unsafe implementation instead? Maybe it would require making it public in CoreLib (but without a contract in CoreFX), which seems reasonable, as that's what we did for RuntimeThread as well.

Comment (mellinoe) [inline src/System.Memory/src/System.Memory.csproj]: I think the idea/goal here is that the "System.Memory" package can be used from any runtime that supports .NET Standard 1.0, which is pretty much every platform anyone cares about. This lets you write a portable library using `Span`, etc. As an optimization, there is a special implementation for `netcoreapp2.0` that happens to be much faster.

File: src/System.Memory/ref/System.Memory.cs (hunk 4 above)
Comment (kouvel) [inline src/System.Memory/ref/System.Memory.cs]: Is this a case where we'd still want to bump the version? If we shipped something at all before (which I'm not sure about), then perhaps the version should be bumped.

Comment (kouvel) [inline src/System.Memory/ref/System.Memory.cs]: Nevermind, apparently this has not shipped before

File: src/System.Memory/src/System.Memory.csproj (hunk 2 above)
Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: Ok makes sense, then I guess we would be shipping the ns1.0 version as well

Comment (ahsonkhan) [inline src/System.Memory/src/System.Memory.csproj]: cc @jkotas

Comment (jkotas) [inline src/System.Memory/src/System.Memory.csproj]: Having the exact same public type it multiple different assemblies tends to cause problems. If we wanted to use the CoreLib Unsafe here, it should be fine but it should:

- CoreLib Unsafe should be given different name or namespace, e.g. Internal.Runtime.CompilerServices.
     or 
- CoreLib Unsafe should be filled in to have all methods that S.R.CS.Unsafe has, and S.R.CS.Unsafe should be turned into façade for netcoreapp20

File: src/System.Memory/src/System/SpanExtensions.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.Memory/src/System/SpanExtensions.cs]: OK. For some reason I thought we had them on many places. As long as all IndexOf methods are in one type, I have no concerns.

File: src/System.Memory/src/System.Memory.csproj (hunk 5 above)
Comment (ahsonkhan) [inline src/System.Memory/src/System.Memory.csproj]: SpanHelpers needs Pinnable. Shall I move both?

File: src/System.Memory/src/System.Memory.csproj (hunk 2 above)
Comment (ahsonkhan) [inline src/System.Memory/src/System.Memory.csproj]: Can this be addressed in a separate PR?

Comment (jkotas) [inline src/System.Memory/src/System.Memory.csproj]: Yes

**Comment by ahsonkhan** [comment]
Waiting for PR #15698 to get merged so that this PR can pass CI.

**Comment by ahsonkhan** [comment]
@kouvel, is this PR good to go?

File: src/System.Memory/src/System.Memory.csproj (hunk 5 above)
Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: Pinnable is only used by the slow span, so I think it needs to be factored out such that it's only included in the slow span assembly.

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

File: src/System.Memory/src/System.Memory.csproj (hunk 2 above)
Comment (joperezr) [inline src/System.Memory/src/System.Memory.csproj]: >@joperezr, is there any reason to build for older frameworks anymore?

To answer your question, yes, if the library is an OOB library (e.g. shipping in its own package, which System.Memory is one of those cases) then we do want to keep old configurations. If it will ship as part of our netcoreapp flat package, then it's ok to remove old configurations

Comment (kouvel) [inline src/System.Memory/src/System.Memory.csproj]: I see, how do I tell whether it's an OOB library or part of the flat package?
